File: Examples/objective-c/ProximityContent/ProximityContent/Estimote/CachingContentFactory.m
//
// Please report any problems with this app template to contact@estimote.com
//

#import "CachingContentFactory.h"

@interface CachingContentFactory ()

@property (nonatomic, readonly) id<BeaconContentFactory> beaconContentFactory;

@property (nonatomic) NSMutableDictionary *cachedContent;

@end

@implementation CachingContentFactory

- (instancetype)initWithBeaconContentFactory:(id<BeaconContentFactory>)beaconContentFactory {
    self = [super init];
    if (self) {
        _beaconContentFactory = beaconContentFactory;
        _cachedContent = [NSMutableDictionary new];
    }
    return self;
}

- (void)contentForBeacon:(CLBeacon *)beacon completion:(void (^)(id content))completion {
    id content = [self.cachedContent objectForKey:beacon.beaconID];
    if (content) {
        completion(content);
    } else {
        [self.beaconContentFactory contentForBeacon:beacon completion:^(id content) {
            self.cachedContent[beacon.beaconID] = content;
            completion(content);
        }];
    }
}

@end

File: Examples/swift/ProximityContent/ProximityContent/Estimote/CachingContentFactory.swift
//
// Please report any problems with this app template to contact@estimote.com
//

class CachingContentFactory: BeaconContentFactory {
    
    private let beaconContentFactory: BeaconContentFactory
    
    private var cachedContent = Dictionary<BeaconID, AnyObject>()
    
    init(beaconContentFactory: BeaconContentFactory) {
        self.beaconContentFactory = beaconContentFactory
    }
    
    func requestContent(for beacon: CLBeacon, completion: @escaping (_ content: AnyObject) -> ()) {
        if let cachedContent: AnyObject = self.cachedContent[beacon.beaconID] {
            completion(cachedContent)
        } else {
            self.beaconContentFactory.requestContent(for: beacon) { (content) in
                self.cachedContent[beacon.beaconID] = content
                completion(content)
            }
        }
    }
    
}
